UnityProject/Assets/Scripts/GameModes/GameMode.cs

```diff
@@ -100,31 +217,80 @@ public bool TrySpawnAntag(PlayerSpawnRequest spawnRequest)
 	/// </summary>
 	protected void SpawnAntag(PlayerSpawnRequest playerSpawnRequest)
 	{
-		if (PossibleAntags.Count > 0)
+		if (PossibleAntags.Count <= 0)
 		{
-			int randIndex = Random.Range(0, PossibleAntags.Count);
-			AntagManager.Instance.ServerSpawnAntag(PossibleAntags[randIndex], playerSpawnRequest);
+			Logger.LogError("PossibleAntags is empty! Game mode must have some if spawning antags.",
+				Category.GameMode);
+			return;
 		}
+
+		int randIndex = Random.Range(0, PossibleAntags.Count);
+		var antag = PossibleAntags[randIndex];
+		if (!AllocateJobsToAntags &&antag.AntagOccupation == null)
+		{
+			Logger.LogErrorFormat("AllocateJobsToAntags is false but {0} AntagOccupation is null! " +
+								  "Game mode must either set AllocateJobsToAntags or possible antags neeed an AntagOccupation.",
+				Category.GameMode, antag.AntagName);
+			return;
+		}
+		AntagManager.Instance.ServerSpawnAntag(PossibleAntags[randIndex], playerSpawnRequest);
 	}
 
 	/// <summary>
-	/// Determine what to do with new players that join
+	/// Start the round
 	/// </summary>
-	// public virtual void SetupNewPlayer()
-	// {
-	// 	if (GameManager.Instance.RoundStarted)
-	// 	{
-	// 		UIManager.Display.SetScreenForJobSelect();
-	// 	}
-	// }
+	public virtual void StartRound()
+	{
+		Logger.LogFormat("Starting {0} round!", Category.GameMode, Name);
+
+		List<PlayerSpawnRequest> playerSpawnRequests;
+		List<PlayerSpawnRequest> antagSpawnRequests;
+		int antagsToSpawn = CalculateAntagCount(PlayerList.Instance.ReadyPlayers.Count);
+		var jobAllocator = new JobAllocator();
+		var playerPool = PlayerList.Instance.ReadyPlayers;
+		if (AllocateJobsToAntags)
+		{
+			// Allocate jobs to all players first then choose antags
+			playerSpawnRequests = jobAllocator.DetermineJobs(playerPool);
+			// TODO: add antag pref checks here
+			var antagCandidates = playerSpawnRequests.Where(p =>
+					!NonAntagJobTypes.Contains(p.RequestedOccupation.JobType));
+			antagSpawnRequests = antagCandidates.PickRandom(antagsToSpawn).ToList();
+			// Player and antag spawn requests are kept separate to stop players being spawned twice
+			playerSpawnRequests.RemoveAll(antagSpawnRequests.Contains);
+		}
+		else
+		{
+			// Choose antags first then allocate jobs to all other players
+			// TODO: add antag pref checks here

```

Comment:
Is this implemented or should an issue be created for this TODO?

Reply:
I didn't read your PR well enough, especially the part where you said that it will be in a separate PR.